### PR TITLE
Add generic application statistics overlay

### PR DESCRIPTION
**Pull request by sergystepanov**
This feature adds an option to gather and display various statistics on the client side.
The statistics module includes some examples of the submodule structure in order to be able to write your own client or server params tracking code.
By default the feature toggle is bound to the `backspace` key.
Display module of the feature uses and depends on the RAF callbacks for smoother animation limited by default to 32fps.

feature003

### CODE REVIEW

**Comment by giongto35** [comment]
Good work ! I 'm really looking for this diff. I also thing I can do more things with canvas as well (like realtime chit chat :-?)

**Comment by sergystepanov** [comment]
> Good work ! I 'm really looking for this diff.

Thanks, hope I'll have time to finish it on the next week (:/.

>  I also thing I can do more things with canvas as well (like realtime chat messaging :-?)

Just why?

**Comment by sergystepanov** [comment]
And this scale stuff with blur-everything have to be dealt with. Literally unplayable otherwise. (:

File: web/js/stats/stats.js [hunk 1]
```diff
@@ -0,0 +1,348 @@
...
+
+        // UI
+        const ui = moduleUi('Ping', true);
+
+        const onPingRequest = (data) => previous = data.time;
+
+        const onPingResponse = () => {
+            length++;
+            const delta = Date.now() - previous;
+            mean += Math.round((delta - mean) / length);
```
Comment (giongto35) [inline web/js/stats/stats.js:199]: oh first time know mean from stream can calculate this way :D . Intersting

File: web/js/stats/stats.js [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+     *  void enable()
+     *  void disable()
+     *  void render()
+     *
+     * @version 1
+     */
+    const clientMemory = (() => {
+        let active = false;
+
+        const ui = moduleUi('Memory', false, 'B');
```
Comment (giongto35) [inline web/js/stats/stats.js:239]: I think you can write it more naturally by provide postfix as a format function with value input.
ui = moduleUi('Memory', false, formatMemory);
ui = moduleUi('Latency', false, formatLatency);

File: web/js/stats/stats.js [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+ * <- HELP_OVERLAY_TOGGLED
+ *
+ * @version 1
+ */
+const stats = (() => {
+    const modules = [];
+    let tempHide = false;
+
+    // internal rendering stuff
+    const drawFps = 32;
```
Comment (giongto35) [inline web/js/stats/stats.js:15]: I think drawFps 10 is enough. Canva Draw is quite heavy I think

File: web/js/stats/stats.js [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+            // 0,0   w,0   0,0   w,0   0,0     w,0
+            // +-------+   +-------+   +---------+
+            // |       |   |+---+  |   |+---+    |
+            // |       |   ||||||  |   ||||||+---+
+            // |       |   ||||||  |   |||||||||||
+            // +-------+   +----+--+   +---------+
+            // 0,h   w,h   0,h   w,h   0,h     w,h
+            // []          [3]         [3, 2]
+            //
+            // O(N+N) :( can be O(1) without visual scale
```
[the remark this answers is not in the document]
Comment (sergystepanov) [reply web/js/stats/stats.js:87]: Just draw over the current one.

File: web/js/stats/stats.js [hunk 5]
```diff
@@ -0,0 +1,348 @@
...
+
+            _context.clearRect(0, 0, _canvas.width, _canvas.height);
+
+            maxN = data[0] || 1;
+            minN = 0;
+            for (let k = 1; k < data.length; k++) {
```
[the remark this answers is not in the document]
Comment (sergystepanov) [reply web/js/stats/stats.js:93]: They could, but for this everything is [slower](https://jsperf.com/apply-vs-loop-for-max/3).

File: web/js/stats/stats.js (hunk 1 above)
Comment (sergystepanov) [reply web/js/stats/stats.js:199]: [Yeah](https://en.wikipedia.org/wiki/Moving_average).

File: web/js/stats/stats.js (hunk 2 above)
Comment (sergystepanov) [reply web/js/stats/stats.js:239]: Mmm... probably. I'll think.

File: web/js/stats/stats.js (hunk 3 above)
Comment (sergystepanov) [reply web/js/stats/stats.js:15]: Not at all. I dropped it from 60 (or your monitor refresh rate) just because it's too distracting.

Comment (sergystepanov) [reply web/js/stats/stats.js:15]: I'll add this value into the settings module which I've started to write.

**Pull request merged**